ingest/clusters.py

```diff
@@ -1,45 +1,77 @@
-import copy
-from typing import Dict
+from typing import Dict, Generator, List, Tuple, Union  # noqa: F401
+from dataclasses import dataclass
+from mypy_extensions import TypedDict
 
-from ingest_files import IngestFiles
+from ingest_files import DataArray
+from annotations import Annotations
 
 
-class Clusters(IngestFiles):
+@dataclass
+class DomainRanges(TypedDict):
+    x: List
+    y: List
+    z: List = None
 
-    ALLOWED_FILE_TYPES = ["text/csv", "text/plain", "text/tab-separated-values"]
-    COLLECTION_NAME = "clusters"
+
+class Clusters(Annotations):
+    # Need to delete after refactor
+    COLLECTION_NAME = "cluster_groups"
     SUBCOLLECTION_NAME = "data"
 
+    ALLOWED_FILE_TYPES = ["text/csv", "text/plain", "text/tab-separated-values"]
+    LINEAR_DATA_TYPE = 'ClusterGroup'
+
+    @dataclass
+    class Document(TypedDict):
+        name: str
+        # 3d or 2d cluster
+        cluster_type: str
+        # Array of Hash objects that describe all extra "annotation" columns
+        cell_annotations: List
+        file_id: str
+        study_id: str
+        # Hash containing min/max arrays for each axis in the cluster plot
+        domain_ranges: DomainRanges = None
+
     def __init__(
         self,
         file_path: str,
-        file_id: str,
-        study_accession: str,
-        *,
+        study_id: str,
+        study_file_id: str,
         name: str,
+        *,
         domain_ranges: Dict = None,
     ):
-        IngestFiles.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
-        self.header = self.get_next_line(increase_line_count=False)
-        # Second line in cluster is metadata_type
-        self.metadata_types = self.get_next_line(increase_line_count=False)
+        Annotations.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
+        self.determine_coordinates_and_cell_names()
         self.source_file_type = "cluster"
-        self.has_z = True if ("z" in self.header or "Z" in self.header) else False
-        self.cell_annotations = self.create_cell_annotations_field()
+        self.cluster_type = (
+            '2d'
```

Comment:
This is backwards - it should be 2d if z/Z is NOT in the headers.

Reply:
91e7310